base/loading.jl

```diff
@@ -1074,10 +1074,12 @@ function _require(pkg::PkgId)
     loading = get(package_locks, pkg, false)
     if loading !== false
         # load already in progress for this module
-        wait(loading)
+        lock(loading) do
+            wait(loading)
```

Comment:
This lock/wait code pattern is always wrong (it is why it forces you to write this out), since it is failing to protect the data. You'll need to rewrite much more of this code to be thread-safe and data-race-free.

Reply:
Can you expand on that? The usage pattern of Threads.Condition is something that is sparsely documented https://docs.julialang.org/en/v1/base/multi-threading/#Synchronization, so if this is "always" wrong it would be great if you could improve the documentation.

Is your point that I might be woken to soon and I need to ensure that the condition I am waiting for has actually been met?

Reply:
When you take a lock, you need to ensure it contains all of the shared data

Reply:
I think we could use a lock for all the global package state, which can be shared by all the per-package Conditions?